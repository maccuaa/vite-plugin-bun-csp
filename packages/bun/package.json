{
  "name": "vite-plugin-bun-csp",
  "version": "0.1.2",
  "description": "A Vite plugin that generates and injects a Content Security Policy (CSP) for your SPA application.",
  "author": "Andrew MacCuaig <andrewmaccuaig@gmail.com>",
  "license": "MIT",
  "types": "./dist/index.d.ts",
  "exports": "./dist/index.js",
  "type": "module",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/maccuaa/vite-plugin-bun-csp.git"
  },
  "keywords": [
    "vite",
    "vite-plugin",
    "csp",
    "content-security-policy",
    "bun",
    "spa"
  ],
  "bugs": {
    "url": "https://github.com/maccuaa/vite-plugin-bun-csp/issues"
  },
  "homepage": "https://github.com/maccuaa/vite-plugin-bun-csp#readme",
  "files": [
    "dist"
  ],
  "peerDependencies": {
    "vite": "^4.0.0 || ^5.0.0 || ^6.0.0"
  },
  "devDependencies": {
    "@tsconfig/bun": "1.0.7",
    "@types/bun": "1.1.16",
    "typescript": "5.5.3",
    "vite": "6.0.3"
  }
}
