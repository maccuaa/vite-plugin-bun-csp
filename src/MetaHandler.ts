import type { ScriptHandler } from "./ScriptHandler";
import type { CSPPolicy } from "./types";

export class MetaHandler implements HTMLRewriterTypes.HTMLRewriterElementContentHandlers {
  private scriptHandler: ScriptHandler;
  private policy: CSPPolicy;

  constructor(scriptHandler: ScriptHandler, policy: CSPPolicy) {
    this.scriptHandler = scriptHandler;
    this.policy = policy;
  }

  element(element: HTMLRewriterTypes.Element) {
    const attributeName = "http-equiv";

    if (!element.hasAttribute(attributeName)) {
      return;
    }

    const attributeValue = element.getAttribute(attributeName);

    if (attributeValue?.toLowerCase() !== "content-security-policy") {
      return;
    }

    const fileHashes = this.scriptHandler.hashValue;

    if (fileHashes) {
      // Add the file hashes to the correct directive. Try script-src-elem first. Fallback to script-src.
      // If no script-src then create and add the hash to the script-src-elem directive.
      if (this.policy["script-src-elem"]) {
        this.policy["script-src-elem"].push(fileHashes);
      } else if (this.policy["script-src"]) {
        this.policy["script-src"].push(fileHashes);
      } else {
        this.policy["script-src-elem"] = [fileHashes];
      }
    }

    const csp = Object.entries(this.policy)
      .map(([name, value]) => `${name} ${value.join(" ")}`)
      .join("; ");

    element.setAttribute("content", csp);
  }
}
