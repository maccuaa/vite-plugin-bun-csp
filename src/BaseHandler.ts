import { type CryptoHasher } from "bun";
import type { AssetCache, HashAlgorithm } from "./types";

export class BaseHandler {
  private algorithm: HashAlgorithm;
  private cache: AssetCache;

  private hasher: CryptoHasher;
  private hashes: string[] = [];

  constructor(algorithm: HashAlgorithm, cache: AssetCache) {
    this.algorithm = algorithm;
    this.hasher = new Bun.CryptoHasher(algorithm);
    this.cache = cache;
  }

  get hashValues() {
    return this.hashes.map((h) => `'${h}'`).join(" ");
  }

  protected calculateHash = (contentsToHash: string) => {
    const hashed = this.hasher.update(contentsToHash).digest("base64");

    const hash = `${this.algorithm}-${hashed}`;

    this.saveHash(hash);

    return hash;
  };

  protected getFileContents = async (filePath: string): Promise<string> => {
    if (filePath.startsWith("https://")) {
      const response = await fetch(filePath);
      return await response.text();
    }

    const fileContents = this.cache.get(filePath);

    if (!fileContents) {
      // https://github.com/oven-sh/bun/issues/15852
      // throw new Error(`${filePath} not found in cache`);
      return "";
    }

    return fileContents;
  };

  protected saveHash = (hash: string) => {
    if (!this.hashes.includes(hash)) {
      this.hashes.push(hash);
    }
  };
}
