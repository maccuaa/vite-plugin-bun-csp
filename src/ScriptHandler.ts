import { type CryptoHasher } from "bun";
import type { AssetCache, HashAlgorithm } from "./types";

export class ScriptHandler implements HTMLRewriterTypes.HTMLRewriterElementContentHandlers {
  private algorithm: HashAlgorithm;
  private cache: AssetCache;

  private hasher: CryptoHasher;
  private hashes: string[] = [];

  private textContent = "";
  private currentElement?: HTMLRewriterTypes.Element;

  constructor(algorithm: HashAlgorithm, cache: AssetCache) {
    this.algorithm = algorithm;
    this.hasher = new Bun.CryptoHasher(algorithm);
    this.cache = cache;
  }

  element(element: HTMLRewriterTypes.Element) {
    const attributeName = "src";

    if (!element.hasAttribute(attributeName)) {
      // Processing an inline script. Wait for the text method to be invoked.
      // 1. Save the element.
      // 2. Reset the text content buffer
      this.currentElement = element;
      this.textContent = "";

      return;
    }

    const filePath = element.getAttribute(attributeName);

    if (!filePath) {
      throw new Error(`${attributeName} attribute does not have a value`);
    }

    const contentsToHash = this.cache.get(filePath);

    if (!contentsToHash) {
      throw new Error(`${filePath} not found in cache`);
    }

    this.calculateHash(contentsToHash);
  }

  text(chunk: HTMLRewriterTypes.Text): void | Promise<void> {
    this.textContent += chunk.text;

    if (chunk.lastInTextNode) {
      this.calculateHash(this.textContent);
    }
  }

  get hashValue() {
    return this.hashes.map((h) => `'${h}'`).join(" ");
  }

  private calculateHash = (contentsToHash: string) => {
    const hashed = this.hasher.update(contentsToHash).digest("base64");

    const cspValue = `${this.algorithm}-${hashed}`;

    this.hashes.push(cspValue);

    if (!this.currentElement) {
      throw new Error("calculateHash method called before element was set");
    }

    this.currentElement.setAttribute("integrity", cspValue);
  };
}
