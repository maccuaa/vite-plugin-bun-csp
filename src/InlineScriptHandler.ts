import { BaseHandler } from "./BaseHandler";

// You use the Subresource Integrity feature by
// specifying a base64-encoded cryptographic hash
// of a resource (file) you're telling the browser
// to fetch, in the value of the integrity attribute
// of a <script> element or a <link> element with
//  rel="stylesheet", rel="preload", or rel="modulepreload".

export class InlineScriptHandler extends BaseHandler implements HTMLRewriterTypes.HTMLRewriterElementContentHandlers {
  private textContent = "";
  private currentElement?: HTMLRewriterTypes.Element;

  async element(element: HTMLRewriterTypes.Element) {
    // Check if we're processing an inline script.
    // If yes, wait for the text method to be invoked so we can hash the text content
    if (element.hasAttribute("src")) {
      return;
    }

    // 1. Save the element.
    // 2. Reset the text content buffer
    this.currentElement = element;
    this.textContent = "";
  }

  text(chunk: HTMLRewriterTypes.Text): void | Promise<void> {
    this.textContent += chunk.text;

    if (chunk.lastInTextNode) {
      const hash = this.calculateHash(this.textContent);

      if (!this.currentElement) {
        throw new Error("text method called before element was set");
      }

      this.currentElement.setAttribute("integrity", hash);
    }
  }
}
