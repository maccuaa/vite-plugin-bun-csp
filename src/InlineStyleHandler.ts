import { BaseHandler } from "./BaseHandler";

export class InlineStyleHandler extends BaseHandler implements HTMLRewriterTypes.HTMLRewriterElementContentHandlers {
  private textContent = "";
  private currentElement?: HTMLRewriterTypes.Element;

  async element(element: HTMLRewriterTypes.Element) {
    // 1. Save the element.
    // 2. Reset the text content buffer
    this.currentElement = element;
    this.textContent = "";
  }

  text(chunk: HTMLRewriterTypes.Text): void | Promise<void> {
    this.textContent += chunk.text;

    if (chunk.lastInTextNode) {
      const hash = this.calculateHash(this.textContent);

      if (!this.currentElement) {
        throw new Error("text method called before element was set");
      }

      this.currentElement.setAttribute("integrity", hash);
    }
  }
}
