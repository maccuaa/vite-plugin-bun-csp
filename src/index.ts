import type { IndexHtmlTransformContext, Plugin } from "vite";
import type { CspPolicy, CspPluginConfiguration } from "./types";
import { ScriptHandler } from "./ScriptHandler";
import { StyleHandler } from "./StyleHandler";
import { InlineStyleHandler } from "./InlineStyleHandler";
import { InlineScriptHandler } from "./InlineScriptHandler";
import { buildCsp, createAssetCache } from "./utils";

/**
 * Default CSP policy.
 * @link https://web.dev/articles/strict-csp
 */
export const DEFAULT_CSP_POLICY: CspPolicy = {
  "base-uri": ["'none'"],
  "default-src": ["'self'"],
  "object-src": ["'none'"],
};

/**
 *
 * @param options
 * @returns
 */
export const generateCspPlugin = (options: CspPluginConfiguration = {}): Plugin => {
  const { algorithm = "sha384" } = options;

  const policy = options.policy ?? { ...DEFAULT_CSP_POLICY };

  return {
    name: "generate-csp",
    enforce: "post",
    apply: "build",

    transformIndexHtml: (html: string, { bundle }: IndexHtmlTransformContext) => {
      const assetCache = createAssetCache(bundle);

      const rewriter = new HTMLRewriter();
      const scriptHandler = new ScriptHandler(algorithm, assetCache);
      const inlineScriptHandler = new InlineScriptHandler(algorithm, assetCache);
      const styleHandler = new StyleHandler(algorithm, assetCache);
      const inlineStyleHandler = new InlineStyleHandler(algorithm, assetCache);

      const newHtml = rewriter
        .on("script", scriptHandler)
        .on("script", inlineScriptHandler)
        .on("link", styleHandler)
        .on("style", inlineStyleHandler)
        .transform(html);

      const csp = buildCsp(policy, { scriptHandler, inlineScriptHandler, styleHandler, inlineStyleHandler });

      const tag = `<meta http-equiv="Content-Security-Policy" content="${csp}" />`;

      if (newHtml.includes("<head>")) {
        return newHtml.replace(/<head>/, `<head>${tag}`);
      } else {
        return newHtml.replace(/(<html.*>)/, `$1<head>${tag}`);
      }
    },
  };
};
