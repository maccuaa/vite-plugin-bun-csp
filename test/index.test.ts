import { describe, expect, it, test } from "bun:test";
import { generateCspPlugin } from "../src";
import type { CspPluginConfiguration, HashAlgorithm } from "../src/types";
import { build } from "vite";
import type { RollupOutput } from "rollup";
import { resolve } from "node:path";
import { ScriptHandler } from "../src/ScriptHandler";
import { StyleHandler } from "../src/StyleHandler";

async function buildVite(entryPath: string, pluginConfig?: CspPluginConfiguration) {
  const projectRoot = resolve(__dirname, entryPath);
  const entryFilename = "index.html";

  const outputs = (await build({
    root: projectRoot,
    logLevel: "error",
    plugins: [generateCspPlugin(pluginConfig)],
  })) as RollupOutput | RollupOutput[];
  const output = Array.isArray(outputs) ? outputs[0].output : outputs.output;

  const file = output.find((item) => item.fileName === entryFilename);

  if (!file) {
    throw new Error(`File with name ${entryFilename} not found in output.`);
  }

  let codeStr = "";
  if ("source" in file) {
    codeStr = file.source.toString();
  } else if ("code" in file) {
    codeStr = file.code;
  } else {
    throw new Error("The file object lacks both `source` and `code` properties.");
  }
  return codeStr;
}

describe("vite-plugin-bun-csp", () => {
  it("should add a default CSP tag to the html", async () => {
    const output = await buildVite("./fixtures/sanity");
    expect(output).toMatchSnapshot();
  });

  describe("scripts", () => {
    const cases = [
      ["inline", "./fixtures/script-inline"],
      ["local", "./fixtures/script-local"],
      ["external", "./fixtures/script-external"],
    ];

    test.each(cases)("it should handle %s scripts", async (_, path) => {
      const output = await buildVite(path);
      expect(output).toMatchSnapshot();
    });
  });

  describe("styles", () => {
    const cases = [
      ["inline", "./fixtures/style-inline"],
      ["local", "./fixtures/style-local"],
      ["external", "./fixtures/style-external"],
    ];

    test.each(cases)("it should handle %s styles", async (_, path) => {
      const output = await buildVite(path);
      expect(output).toMatchSnapshot();
    });

    it("should automatically handle Google Fonts", async () => {
      const output = await buildVite("./fixtures/google-fonts");
      expect(output).toMatchSnapshot();
    });
  });

  describe("edge-cases", () => {
    it("should skip elements that already have an integrity hash", () => {
      const cache = new Map<string, string>();

      const rewriter = new HTMLRewriter();
      const scriptHandler = new ScriptHandler("sha256", cache);
      const styleHandler = new StyleHandler("sha256", cache);

      const html = `
        <html>
          <head>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" />
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      `;

      const newHtml = rewriter.on("script", scriptHandler).on("link", styleHandler).transform(html);

      expect(newHtml).toMatchSnapshot();
      expect(scriptHandler.hashValues).toEqual(
        "'sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh'"
      );
      expect(styleHandler.hashValues).toEqual(
        "'sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa'"
      );
    });

    it("should skip link elements that are missing attributes", () => {
      const cache = new Map<string, string>();

      const rewriter = new HTMLRewriter();
      const styleHandler = new StyleHandler("sha256", cache);

      const html = `
        <html>
          <head>
            <!-- rel is not "stylesheet" -->
            <link rel="icon" href="favicon.ico" />
            <!-- element has no rel attribute -->
            <link href="https://vite.dev" />
            <!-- element has no href -->
            <link rel="stylesheet" />
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      `;

      const newHtml = rewriter.on("link", styleHandler).transform(html);

      expect(newHtml).toEqual(html);
    });

    it("should not duplicate hash or url values", () => {
      const cache = new Map<string, string>();

      const rewriter = new HTMLRewriter();
      const scriptHandler = new ScriptHandler("sha384", cache);
      const styleHandler = new StyleHandler("sha384", cache);

      const html = `
        <html>
          <head>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" />
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" />
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      `;

      const newHtml = rewriter.on("script", scriptHandler).on("link", styleHandler).transform(html);

      expect(newHtml).toMatchSnapshot();
      expect(scriptHandler.hashValues).toEqual(
        "'sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh'"
      );
      expect(styleHandler.hashValues).toEqual(
        "'sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa'"
      );
      expect(styleHandler.urls).toEqual("https://cdn.jsdelivr.net");
    });

    it("should throw if the file contents are not found in the cache", () => {
      const cache = new Map<string, string>();

      const rewriter = new HTMLRewriter();
      const scriptHandler = new ScriptHandler("sha256", cache);
      const styleHandler = new StyleHandler("sha256", cache);

      const html = `
        <html>
          <head>
            <link href="style.css" rel="stylesheet" />
            <script src="index.js></script>
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      `;

      const newHtml = rewriter.on("script", scriptHandler).on("link", styleHandler).transform(html);

      expect(newHtml).toEqual(html);
    });
  });

  describe("algorithms", () => {
    const cases: HashAlgorithm[] = ["sha256", "sha384", "sha512"];

    test.each(cases)("should support the %s algorithm", async (algorithm) => {
      const output = await buildVite(`./fixtures/${algorithm}`, { algorithm });
      expect(output).toMatchSnapshot();
    });
  });

  describe("policy", () => {
    it("should add to existing policy", async () => {
      const output = await buildVite("./fixtures/policy", {
        policy: {
          "font-src": ["fonts.gstatic.com"],
          "style-src": ["fonts.googleapis.com"],
        },
      });
      expect(output).toMatchSnapshot();
    });
  });
});
