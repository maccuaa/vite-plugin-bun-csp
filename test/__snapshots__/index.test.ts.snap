// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`vite-plugin-bun-csp scripts it should handle inline scripts 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha384-38HOgRVS9paAKT7chxOOnubobWZ3kJfici4Bp1PHnTX8gK1inuKN/7g2KVNbcmuY'; script-src 'strict-dynamic'; style-src 'self'" /> </head>
  <body>
    <h1 id="app"></h1>
    <script>
      document.querySelector("#app").innerHTML = \`Hello World\`;
    </script>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp scripts it should handle local scripts 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh'; script-src 'strict-dynamic'" />   <script type="module" crossorigin src="/assets/index-DXh1w3r_.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
 </head>
  <body>
    <h1 id="app"></h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp scripts it should handle external scripts 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha384-DpVxUeeBWjUvUV1czyIHJAjh+jYUZFu2lLakbdua5vbwOrBGi1UgaKCHjTC+x3Ky' 'sha384-kgS1YDRuhk0bciElGkbKuHKqpmzV1kVKPcsiQlZ2/UrCfdNkc0GFx6OFgu5CKekd'; script-src 'strict-dynamic'; style-src 'self'" />
    <script src="https://cdn.jsdelivr.net/npm/dayjs@1.11.13/dayjs.min.js" integrity="sha384-DpVxUeeBWjUvUV1czyIHJAjh+jYUZFu2lLakbdua5vbwOrBGi1UgaKCHjTC+x3Ky" crossorigin="anonymous"></script>
  </head>
  <body>
    <h1>Hello World</h1>
    <h3 id="date"></h3>
    <script>
      document.querySelector("#date").innerHTML = dayjs().format("DD-MM-YYYY");
    </script>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp styles it should handle inline styles 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; style-src 'sha384-fw1AlJQQH9oKSLUz3uHsWxT8diqDf9FgPL3MrSOM+JAV6LQSoPIrMUVTdfYpnZxg'" />
    <style>
      h1 {
        color: purple;
      }
    </style>
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp styles it should handle local styles 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'" />
    <link rel="stylesheet" crossorigin href="/assets/index-WbHkZAB3.css" integrity="sha384-+AelCtrokKIdKrnBX1EthqSg5qtANYIjPwDmJidNEJBd65A3+eh+W1t8LgQAf7b9">
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp styles it should handle external styles 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; style-src https://cdn.jsdelivr.net" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" crossorigin="anonymous" />
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp styles should automatically handle Google Fonts 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; style-src 'sha384-bVH4IxTsxUcs8AJWz53Ymr+Vn3ABxlZereLqJsu/ReoAsaaYqN7Hm0+TPFeKhdxJ' https://fonts.googleapis.com; font-src fonts.gstatic.com" />
    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link href="https://fonts.googleapis.com/css2?family=Tomorrow&display=swap" rel="stylesheet" crossorigin="anonymous" />
    <style>
      :root {
        font-family: "Tomorrow";
      }
    </style>
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp edge-cases should skip elements that already have an integrity hash 1`] = `
"
        <html>
          <head>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" />
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      "
`;

exports[`vite-plugin-bun-csp edge-cases should not duplicate hash or url values 1`] = `
"
        <html>
          <head>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <script src="index.js" integrity="sha384-vTb4ka/HRdvMTCK99IHf4grsm4H7ngk8QotM6VmrozV6fFTs6kVHItFSLXWTgnAh"></script>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" />
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/reset-css@5.0.2/reset.min.css" integrity="sha384-keU0jZorLpCR/4h2i+vCgG3TUPKXXt1zOsXoIcwy0uRJKOBQ26zPqJpJj6hPdgpa" crossorigin="anonymous" />
          </head>
          <body>
            <h1 id="app"></h1>
          </body>
        </html>
      "
`;

exports[`vite-plugin-bun-csp algorithms should support the sha256 algorithm 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha256-nP25Pzivzy0Har7NZtMr/TODzfGWdlTrwmomYF2vQXM=' 'sha256-LL1trj6s1DR6qNh4ukIlQj/8NFdHwDQaUeIu87TwMHY=' 'sha256-QuP42mDRshb0EzsB1V3VXyvcvRbVgbx4ir8D/7BFgw0='; style-src 'sha256-LtGmiPOYFkJYXPu3Xt2sSbdUbKJ3U/7Cp0uFAlTEmrU=' https://cdn.jsdelivr.net 'self'; script-src 'strict-dynamic'" />
    <script src="https://cdn.jsdelivr.net/npm/dayjs@1.11.13/dayjs.min.js" integrity="sha256-nP25Pzivzy0Har7NZtMr/TODzfGWdlTrwmomYF2vQXM=" crossorigin="anonymous"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.0.6/css/pico.min.css" integrity="sha256-3V/VWRr9ge4h3MEXrYXAFNw/HxncLXt9EB6grMKSdMI=" crossorigin="anonymous" />
    <style>
      #app {
        color: purple;
      }
    </style>
    <script type="module" crossorigin src="/assets/index-C9hA1kMn.js" integrity="sha256-LL1trj6s1DR6qNh4ukIlQj/8NFdHwDQaUeIu87TwMHY="></script>
    <link rel="stylesheet" crossorigin href="/assets/index-CZz58IoC.css" integrity="sha256-ZLE5urGkSEdDRpzewM/O6m+FePTt7sNKfTrrEy8fT/0=">
  </head>
  <body>
    <main class="container">
      <h1 id="app"></h1>
      <h3 id="date"></h3>
    </main>
    <script>
      document.querySelector("#app").innerHTML = \`Hello World\`;
    </script>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp algorithms should support the sha384 algorithm 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha384-DpVxUeeBWjUvUV1czyIHJAjh+jYUZFu2lLakbdua5vbwOrBGi1UgaKCHjTC+x3Ky' 'sha384-eq2PlAE9Ci0XUAfs71G6Jn4YdOiGl6dhZYvSq5r8okC0tTsxa0kqvpKkyFaDT+kX' 'sha384-38HOgRVS9paAKT7chxOOnubobWZ3kJfici4Bp1PHnTX8gK1inuKN/7g2KVNbcmuY'; style-src 'sha384-2IH9S6CFnaEWZJOjFu078HQpjmJLXFirDvUII2HOVzcVuoXg8rpY+1vtY29CsbPC' https://cdn.jsdelivr.net 'self'; script-src 'strict-dynamic'" />
    <script src="https://cdn.jsdelivr.net/npm/dayjs@1.11.13/dayjs.min.js" integrity="sha384-DpVxUeeBWjUvUV1czyIHJAjh+jYUZFu2lLakbdua5vbwOrBGi1UgaKCHjTC+x3Ky" crossorigin="anonymous"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.0.6/css/pico.min.css" integrity="sha384-7P0NVe9LPDbUCAF+fH2R8Egwz1uqNH83Ns/bfJY0fN2XCDBMUI2S9gGzIOIRBKsA" crossorigin="anonymous" />
    <style>
      #app {
        color: purple;
      }
    </style>
    <script type="module" crossorigin src="/assets/index-C9hA1kMn.js" integrity="sha384-eq2PlAE9Ci0XUAfs71G6Jn4YdOiGl6dhZYvSq5r8okC0tTsxa0kqvpKkyFaDT+kX"></script>
    <link rel="stylesheet" crossorigin href="/assets/index-CZz58IoC.css" integrity="sha384-x7j0GfbBZPnTbD+B5g/PGZOzdW8D2ToXR9FSI/uh+JHDtiuO42/bd5kwlV4UanU2">
  </head>
  <body>
    <main class="container">
      <h1 id="app"></h1>
      <h3 id="date"></h3>
    </main>
    <script>
      document.querySelector("#app").innerHTML = \`Hello World\`;
    </script>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp algorithms should support the sha512 algorithm 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'; script-src-elem 'sha512-FwNWaxyfy2XlEINoSnZh1JQ5TRRtGow0D6XcmAWmYCRgvqOUTnzCxPc9uF35u5ZEpirk1uhlPVA19tflhvnW1g==' 'sha512-n4EoMAP6D7cUYgwfeok3doRXYciju7iuVsvWTL47MUrkr/KYYeTziYpaF196m8ZbIAgzQigkqO4/gRSEs25IqA==' 'sha512-YWgALCmz5eSTVsPOi0xlUmW7XVImjghbsKn0sUTm+RAr7sQfeF4GVRm/SJ+l99sQoluJIM9F+IfG60ufRg3P2w=='; style-src 'sha512-ke/umhQiERl6gm3X5tJlOs/zyCG3VpCXmnMLt24pfv9/Sx3CTcxpHC7BRzBmgD7mE8uR6WVj1TKZvjrYcK0fMw==' https://cdn.jsdelivr.net 'self'; script-src 'strict-dynamic'" />
    <script src="https://cdn.jsdelivr.net/npm/dayjs@1.11.13/dayjs.min.js" integrity="sha512-FwNWaxyfy2XlEINoSnZh1JQ5TRRtGow0D6XcmAWmYCRgvqOUTnzCxPc9uF35u5ZEpirk1uhlPVA19tflhvnW1g==" crossorigin="anonymous"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@picocss/pico@2.0.6/css/pico.min.css" integrity="sha512-UXfikgakSZBii5lkvmDCRO+IYWQhTtwMOJ+3EmGEA+oA82kvbSskgw3OI16Jx1kINgF8aqOkYE+c9h4m6muONg==" crossorigin="anonymous" />
    <style>
      #app {
        color: purple;
      }
    </style>
    <script type="module" crossorigin src="/assets/index-C9hA1kMn.js" integrity="sha512-n4EoMAP6D7cUYgwfeok3doRXYciju7iuVsvWTL47MUrkr/KYYeTziYpaF196m8ZbIAgzQigkqO4/gRSEs25IqA=="></script>
    <link rel="stylesheet" crossorigin href="/assets/index-CZz58IoC.css" integrity="sha512-cdw8JfSRqdKqxBVJa4Qs25AtMyj3BfWW4pKNRUkv+vKTqux0xw5q62UYJlmlyYKUeCdrFo/Vw1KcMsu+2pbF7A==">
  </head>
  <body>
    <main class="container">
      <h1 id="app"></h1>
      <h3 id="date"></h3>
    </main>
    <script>
      document.querySelector("#app").innerHTML = \`Hello World\`;
    </script>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp policy should add to existing policy 1`] = `
"<html>
  <head><meta http-equiv="Content-Security-Policy" content="font-src fonts.gstatic.com; style-src fonts.googleapis.com 'sha384-bVH4IxTsxUcs8AJWz53Ymr+Vn3ABxlZereLqJsu/ReoAsaaYqN7Hm0+TPFeKhdxJ' https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link href="https://fonts.googleapis.com/css2?family=Tomorrow&display=swap" rel="stylesheet" crossorigin="anonymous" />
    <style>
      :root {
        font-family: "Tomorrow";
      }
    </style>
  </head>
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;

exports[`vite-plugin-bun-csp should add a default CSP tag to the html 1`] = `
"<html><head><meta http-equiv="Content-Security-Policy" content="base-uri 'none'; default-src 'self'; object-src 'none'" />
  <body>
    <h1>Hello World</h1>
  </body>
</html>
"
`;
